rfc/006-semantic-versioning.md

```diff
@@ -0,0 +1,76 @@
+2. Node Operators, those running node infrastructure, are predominantly concerned with downtime, complexity and frequency of upgrading and loss of data. They may be also concerned about changes that may break the scripts and tooling they use to supervise their nodes.
+3. External clients, for example, wallets and block explorers are those that perform any of the following:
+     - consume the RPC endpoints of nodes like `/block`
+     - subscribe to events that are streamed via websockets
+     - make queries to the indexer
+  
+    This set are concerned with chain upgrades which will impact their ability to query state and block data as well as broadcast transactions.
+  
+4. There is also a fourth emerging user: IBC module and relayers (although in many ways relayers can also be considered as external clients). The developers of IBC and consumers of their software are concerned about changes that may affect a chains ability to send arbitrary messages to another chain. Specifically, this can happen with a breaking change to the light client verification algorithm.  
+
+In a more broader sense, these four users are inextricably linked and their concerns reflect that of the end user of the applications. The major conclusion that this RFC makes is that **the ability for chains to provide continual service is more important than the actual upgrade burden put on the developers of these chains**. What that means is that although it may be burdensome for application developers to upgrade Tendermint versions, this is far outweighed by the pain node operators experience in upgrading, by the pain that external clients have in accommodating changes and ultimately the pain of end users who are benefiting from the application. Developers have all the time in world to make these changes whilst for other user groups, upgrading can be considered part of this real-time critical path.  
+
+## Proposal
+
+Tendermint will follow a same formatting structure of [SemVer](https://semver.org/) with a major, minor and patch version but with slight adjustments to what they encompass:
+
+A major version in Tendermint guarantees that for the entire cycle of that major version:
+
+- All blocks and state data in a blockchain can be queried. All headers can be verified even across minor version changes. Nodes can both block sync and state sync from genesis to the head of the chain.
+- Nodes in a network are able to communicate and perform BFT state machine replication so long as the agreed network version is the lowest of all nodes in a network. For example, nodes using version 1.5.x and 1.2.x can operate together so long as the network version is 1.2 or lower (but still within the 1.x range). This rule essentially captures the concept of network backwards compatibility.
```

Comment:
> Is "network version" different from "Tendermint version"? It seems like maybe it's different but related?

Network version is a subset of the Tendermint version that pertains to changes made to the p2p layer that would affect communication over the wire.

What I am trying to get at is that we don't provide forward compatibility of both the wire format and data structures, thus if a network agrees to bump the version, those that haven't upgraded will stop. Does that answer your question?

Reply:
I think it does answer my question. I'm trying to think about how to word this promise in a way that would avert my original confusion. Perhaps:

> All nodes in a network operating at or above the agreed network version are able to communicate and perform BFT state replication.

Trying to get at the idea that the whole network doesn't stop just because one node is recalcitrant -- only the laggards get stuck. Maybe there's a better way to put that.